compiler/optimizer/VPHandlers.cpp

```diff
@@ -144,15 +143,14 @@ static void checkForNonNegativeAndOverflowProperties(OMR::ValuePropagation *vp,
          if (high <= 0)
             node->setIsNonPositive(true);
 
-         ///if ((node->getOpCode().isArithmetic() || node->getOpCode().isLoad()) &&
-         ///      ((low > TR::getMinSigned<TR::Int32>()) || (high < TR::getMaxSigned<TR::Int32>())))
-         if ((node->getOpCode().isLoad() &&
-               ((low > TR::getMinSigned<TR::Int32>()) || (high < TR::getMaxSigned<TR::Int32>()))) ||
-               (node->getOpCode().isArithmetic() &&
-                (range->canOverflow() != TR_yes)))
+         if (node->getOpCode().isLoad() &&
+               ((low > TR::getMinSigned<TR::Int32>()) || (high < TR::getMaxSigned<TR::Int32>())))
             {
             node->setCannotOverflow(true);
-            //dumpOptDetails(vp->comp(), "Node %p cannot overflow\n", node);
+            }
+         else if (node->getOpCode().isArithmetic())
+            {
+            node->setCannotOverflow(false);
```

Comment:
Although this change will definitely avoid the problem, I think it's going to be too heavy-handed.  Consider this example:

```
public class Experiment {
    public static final int sub(int i, int j) {
        if (i > 0 && i < 100 && j > 0 && j < 100) {
            return i*j + i/j;
        }
        return 0;
    }

    public static final void main(String[] args) {
        sub(1, 2);
        sub(1, 2);
    }
}
```

Global Value Propagation is able to determine the ranges of the arithmetic operations, and it ought to be able to determine from that that they cannot overflow.

```
   n32n idiv gets new constraint: value 5 is (0 to 99)I
   n29n imul gets new constraint: value 6 is (1 to 9801)I
   n34n iadd gets new constraint: value 7 is (1 to 9900)I
```

The existing code here is testing whether `range->canOverflow() != TR_yes`.  I'm wondering whether the constraint that's associated with the `imul` can have its `range->canOverflow()` return `TR_yes`.  I think the fix will likely be a bit more involved to make that happen.